fish-rust/src/abbrs.rs

```diff
@@ -0,0 +1,479 @@
+    // whether we are derived from a universal variable.
+    pub fn new(
+        name: WString,
+        key: WString,
+        replacement: WString,
+        position: Position,
+        from_universal: bool,
+    ) -> Self {
+        Self {
+            name,
+            key,
+            regex: None,
+            replacement,
+            replacement_is_function: false,
+            position,
+            set_cursor_marker: None,
+            from_universal,
+        }
+    }
+
+    // \return true if this is a regex abbreviation.
+    pub fn is_regex(&self) -> bool {
+        self.regex.is_some()
+    }
+
+    // \return true if we match a token at a given position.
+    pub fn matches(&self, token: &wstr, position: Position) -> bool {
+        if !self.matches_position(position) {
+            return false;
+        }
+        self.regex
+            .as_ref()
+            .map(|r| r.matches_ffi(&token.to_ffi()))
+            .unwrap_or(self.key == token)
+    }
+
+    // \return if we expand in a given position.
+    fn matches_position(&self, position: Position) -> bool {
+        return self.position == Position::Anywhere || self.position == position;
+    }
+}
+
+/// The result of an abbreviation expansion.
+pub struct Replacer {
+    /// The string to use to replace the incoming token, either literal or as a function name.
+    replacement: WString,
+
+    /// If true, treat 'replacement' as the name of a function.
+    is_function: bool,
+
+    /// If set, the cursor should be moved to the first instance of this string in the expansion.
+    set_cursor_marker: Option<WString>,
+}
+
+impl From<Replacer> for abbrs_replacer_t {
+    fn from(value: Replacer) -> Self {
+        let has_cursor_marker = value.set_cursor_marker.is_some();
+        Self {
+            replacement: value.replacement.to_ffi(),
+            is_function: value.is_function,
+            set_cursor_marker: value.set_cursor_marker.unwrap_or_default().to_ffi(),
+            has_cursor_marker,
+        }
+    }
+}
+
+struct Replacement {
+    /// The original range of the token in the command line.
+    range: SourceRange,
+
+    /// The string to replace with.
+    text: WString,
+
+    /// The new cursor location, or none to use the default.
+    /// This is relative to the original range.
+    cursor: Option<usize>,
+}
+
+impl Replacement {
+    /// Construct a replacement from a replacer.
+    /// The \p range is the range of the text matched by the replacer in the command line.
+    /// The text is passed in separately as it may be the output of the replacer's function.
+    fn from(range: SourceRange, mut text: WString, set_cursor_marker: Option<WString>) -> Self {
+        let mut cursor = None;
+        if let Some(set_cursor_marker) = set_cursor_marker {
+            let matched = text
+                .as_char_slice()
+                .windows(set_cursor_marker.len())
+                .enumerate()
```

Comment:
you can use `.position()` instead of `enumerate()`

Reply:
Nice, did not know about position